File: programmers/LV2/시소짝꿍.swift
import Foundation

func combination(_ n: Int) -> Int64 {
    return Int64(n * (n - 1) / 2)
}

func solution(_ weights: [Int]) -> Int64 {
    var answer: Int64 = 0
    
    var weightDic: [Int: Int] = [:]
    for weight in weights {
        if let _ = weightDic[weight] {
            weightDic[weight]! += 1
        } else {
            weightDic[weight] = 1
        }
    }
    
    for weight in weightDic {
        let currentValue = weight.value
        let currentKey = weight.key
        
        // 같은 몸무게가 2명 이상인 경우
        if currentValue >= 2 {
            answer += combination(currentValue)
        }
        
        // 2:3, 2:4, 3:4 비율을 만족하는 몸무게가 있는지?
        if let r1 = weightDic[Int(currentKey / 2) * 3] {
            if currentKey % 2 == 0 {
                answer += Int64(r1 * currentValue)
            }
        }
        
        if let r2 = weightDic[currentKey * 2] {
            answer += Int64(r2 * currentValue)
        }
        
        if let r3 = weightDic[Int(currentKey / 3) * 4] {
            if currentKey % 3 == 0 {
                answer += Int64(r3 * currentValue)
            }
        }
    }
    
    
    return answer
}

File: programmers/LV2/시소 짝꿍.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

long long combination(long long n) {
    return (long long)(n * (n - 1) / 2);
}

long long solution(vector<int> weights) {
    map<int, long long> m;
    long long answer = 0;
    
    for(int i = 0; i < weights.size(); i++) {
        m[weights[i]] += 1;
    }
    
    for(auto i = m.begin(); i != m.end(); i++) {
        int key = i -> first;
        long long value = i -> second;
        
        if(value >= 2) {
            answer += combination(value);
        }
        
        if(key % 2 == 0 && m[key / 2 * 3] != 0) {
            answer += value * m[key * 3 / 2];
        }
        
        if(key % 3 == 0 && m[key / 3 * 4] != 0) {
            answer += value * m[key * 4 / 3];
        }
        
        if(m[key * 2] != 0) {
            answer += value * m[key * 2];
        }
    }
    
    return answer;
}